2.3-TapTree.ipynb

```diff
@@ -0,0 +1,500 @@
+    "    * `[Script Inclusion Proof]` \n",
+    "        * `n x 32Bytes`\n",
+    "\n",
+    "\n",
+    "_Note: As of writing, the Taproot BIP has been updated to reflect 32B public keys in the controlblock. Following examples and exercises continue to use 33B public keys until the taproot Bitcoin Core branch has been updated._\n",
+    "\n",
+    "![test](images/TapTree4.jpg)\n",
+    "\n",
+    "\n",
+    "**Generating the Controlblock**\n",
+    "\n",
+    "We use the the TapTree construct method to generate the taproot output, tweak and controlblocks for all TapScripts.\n",
+    "\n",
+    "**`TapTree.construct()` returns the tuple:**\n",
+    "* `taproot_output_script`, `tweak`, `control_block_map`\n",
+    "* Control block map has key-value pairs: `tapscript` - `controlblock`\n",
+    "        "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Programming Exercise 2.1 - Constructing a Taptree output.\n",
+    "\n",
+    "In the following exercise, please construct the output and segwit address for the TapTree structure shown in section 1C with the huffman TapTree constructor. Please generate new keys for the internal key and pay-to-pubkey TapScripts.\n",
+    "\n",
+    "We will cover TapScripts in more depth in the next chapter, but consider the following method to generate pay-to-pubkey TapScripts:\n",
+    "* `TapLeaf.construct_pk(public_key)`"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Generate Keypairs for internal pubkey and pay-to-pubkey TapScripts.\n",
+    "privkey_internal = ECKey()\n",
+    "privkey_internal.generate()\n",
+    "pubkey_internal = privkey_internal.get_pubkey()\n",
+    "\n",
+    "privkeyA = ECKey()\n",
+    "privkeyB = ECKey()\n",
+    "privkeyC = ECKey()\n",
+    "privkeyD = ECKey()\n",
+    "privkeyA.generate()\n",
+    "privkeyB.generate()\n",
+    "privkeyC.generate()\n",
+    "privkeyD.generate()\n",
+    "pubkeyA = privkeyA.get_pubkey()\n",
+    "pubkeyB = privkeyB.get_pubkey()\n",
+    "pubkeyC = privkeyC.get_pubkey()\n",
+    "pubkeyD = privkeyD.get_pubkey()\n",
+    "\n",
+    "# Construct Pay-toPubkey TapLeafs and Taptree.\n",
+    "TapLeafA = TapLeaf()\n",
+    "TapLeafB = TapLeaf()\n",
+    "TapLeafC = TapLeaf()\n",
+    "TapLeafD = TapLeaf()\n",
+    "TapLeafA.construct_pk(pubkeyA)\n",
+    "TapLeafB.construct_pk(pubkeyB)\n",
+    "TapLeafC.construct_pk(pubkeyC)\n",
+    "TapLeafD.construct_pk(pubkeyD)\n",
+    "\n",
+    "taptree = TapTree()\n",
+    "taptree.key = pubkey_internal\n",
+    "taptree.huffman_constructor([(1,TapLeafA),(1, TapLeafB),(2, TapLeafC),(3,TapLeafD)])\n",
+    "\n",
+    "# Generate taproot output and segwit address.\n",
+    "taproot_script, tweak, control_map = taptree.construct()\n",
+    "pubkey_taproot_bytes = bytes(taproot_script[2:])\n",
+    "segwit_address = program_to_witness(1, pubkey_taproot_bytes)\n",
```

Comment:
This line and the previous line are a little awkward. Seems like we pull out the version and then hard code it back in? If that’s right and theres a better way lets tweak. Otherwise, its fine.

Reply:
Wording is a little awkward, but the steps look ok to me. I've updated the comment.